components/wmtk_components/mesh_decimation/wmtk/components/mesh_decimation/internal/MeshDecimation.cpp

```diff
@@ -0,0 +1,172 @@
+                acc_edge.scalar_attribute(m_mesh.switch_tuples(face, {PV, PE})) = 1;
+            }
+        }
+        break;
+    }
+    default:
+        std::runtime_error("MeshDecimation.cpp: mesh_decimation component only supports tetmesh "
+                           "and trimesh for now!");
+    }
+
+    for (const Tuple& edge : m_mesh.get_all(PE)) {
+        if (acc_vertex.scalar_attribute(edge) == 1 ||
+            acc_vertex.scalar_attribute(m_mesh.switch_tuple(edge, PV)) == 1) {
+            acc_edge.scalar_attribute(edge) = 1;
+        }
+        acc_len.scalar_attribute(edge) = (acc_pos.vector_attribute(edge) -
+                                          acc_pos.vector_attribute(m_mesh.switch_tuple(edge, PV)))
+                                             .norm();
+    }
+
+    auto op_scaffold = std::make_shared<operations::EdgeCollapse>(m_mesh);
+
+    op_scaffold->add_invariant(
+        std::make_shared<TodoInvariant>(m_mesh, edge_handle.as<int64_t>(), 0));
+    op_scaffold->add_invariant(std::make_shared<TodoSmallerInvariant>(
+        m_mesh,
+        edge_len_handle.as<double>(),
+        4.0 / 5.0 * m_target_len));
+
+    auto m_amips = std::make_shared<function::AMIPS>(m_mesh, position);
+    auto m_link_conditions = std::make_shared<InvariantCollection>(m_mesh);
+    m_link_conditions->add(std::make_shared<MultiMeshLinkConditionInvariant>(m_mesh));
+    auto m_function_invariant =
+        std::make_shared<SmallerFunctionInvariant>(m_mesh.top_simplex_type(), m_amips, 30);
+    auto m_inversion_invariant =
+        std::make_shared<SimplexInversionInvariant>(m_mesh, position.as<double>());
+
+    // Edge length update
+    auto compute_edge_length = [](const Eigen::MatrixXd& P) -> Eigen::VectorXd {
+        assert(P.cols() == 2);
+        assert(P.rows() == 2 || P.rows() == 3);
+        return Eigen::VectorXd::Constant(1, (P.col(0) - P.col(1)).norm());
+    };
+    auto m_edge_length_update =
+        std::make_shared<wmtk::operations::SingleAttributeTransferStrategy<double, double>>(
+            edge_len_handle,
+            position,
+            compute_edge_length);
+    auto m_prio_short_edges_first = [&](const simplex::Simplex& s) {
+        assert(s.primitive_type() == PrimitiveType::Edge);
+        auto acc = m_mesh.create_accessor<double>(edge_len_handle);
+        return std::vector<double>({acc.scalar_attribute(s.tuple())});
+    };
+
+
+    op_scaffold->add_invariant(m_link_conditions);
+    op_scaffold->add_invariant(m_function_invariant);
+    op_scaffold->add_invariant(m_inversion_invariant);
+
+    op_scaffold->add_transfer_strategy(m_edge_length_update);
+
+    op_scaffold->set_priority(m_prio_short_edges_first);
+
+    op_scaffold->set_new_attribute_strategy(
+        position,
+        wmtk::operations::CollapseBasicStrategy::Mean);
+    op_scaffold->set_new_attribute_strategy(vertex_handle);
+
+    op_scaffold->set_new_attribute_strategy(edge_handle);
```

Comment:
You need to update your edge_todos correctly. I think the best way to do it is to use an attribute transfer and compute the edge_todo from the vertex tag attribute.